distributed_shampoo/shampoo_fsdp_utils.py

```diff
@@ -0,0 +1,900 @@
+
+    Starting from the leftmost dimension, the largest possible slices in each dimension
+    (with the remaining dimensions on the right retaining the original shape) are split off.
+
+    2D example:
+     _______________                  _______________
+    |       ________|                |       ________|
+    |______|        |                |______|________|
+    |               |                |               |
+    |    shard  ____|       ->       |_______________|
+    |__________|    |                |__________|    |
+    |               |                |               |
+    |_______________|                |_______________|
+
+    Args:
+        tensor (Tensor): Flattened gradient or tensor to split.
+        orig_shape (torch.Size): Shape of original tensor that tensor is a slice of.
+        start_idx (int): Flattened index in original tensor where tensor starts.
+        end_idx (int): Flattened index in original tensor where tensor ends (inclusive).
+        include_placeholder (bool): Whether to include empty partitioned tensors as placeholders. (Default: False)
+
+    Returns:
+        split_tensors (List[Tensor]): List of tensors.
+
+    """
+    if len(tensor.size()) != 1:
+        logger.info(
+            f"Input tensor is not flat, has shape {tensor.size()}. Continuing without splitting."
+        )
+        return tensor
+
+    end_idx += 1  # correct off-by-one (FSDP shard_param_info provides inclusive index)
+    assert (
+        end_idx - start_idx == tensor.size()[0]
+    ), f"Start/end indices do not match tensor size: start {start_idx} end {end_idx}, tensor size {tensor.size()}"
+
+    # maintains the order that partitions had in the flat tensor
+    split_tensors_left = []
+    split_tensors_right = []
+    left_idx = None
+    right_idx = None
+    center_partition = False
+    for i in range(1, len(orig_shape) + 1):
+        remaining_size = prod(orig_shape[i:])
+        left_idx_new = int(np.ceil(start_idx / remaining_size)) * remaining_size
+        right_idx_new = end_idx // remaining_size * remaining_size  # floor
+
+        # first iteration (largest convex partition in the center)
+        if not center_partition:
+            if left_idx_new <= right_idx_new:
+                if left_idx_new < right_idx_new:
+                    split_tensors_left.append(
+                        torch.narrow(
+                            tensor,
+                            0,
+                            left_idx_new - start_idx,
+                            right_idx_new - left_idx_new,
+                        )
+                        .view([-1] + list(orig_shape[i:]))
+                        .squeeze()
+                    )
+                elif include_placeholder:
+                    split_tensors_left.append(torch.tensor([]).to(tensor.device))
+                left_idx = left_idx_new
+                right_idx = right_idx_new
+                center_partition = True
+            continue
```

Comment:
Let's add some spacing and comments here. Would be good to note that the center partition is included in the `split_tensors_left` list.
```suggestion
        # handle largest convex partition in the center
        if not center_partition:
            if left_idx_new <= right_idx_new:
                if left_idx_new < right_idx_new:
                    split_tensors_left.append(
                        torch.narrow(
                            tensor,
                            0,
                            left_idx_new - start_idx,
                            right_idx_new - left_idx_new,
                        )
                        .view([-1] + list(orig_shape[i:]))
                        .squeeze()
                    )
                
                # stores empty tensor if including placeholder
                elif include_placeholder:
                    split_tensors_left.append(torch.tensor([]).to(tensor.device))
                
                # update left and right idx
                left_idx = left_idx_new
                right_idx = right_idx_new
                center_partition = True
                
            continue
```

Reply:
I didn't think it would be that important, since the left and right lists are joined at the end anyways 😆